SSA/Core/EffectKind.lean

```diff
@@ -0,0 +1,208 @@
+  cases e1 <;> cases e2 <;> cases e3 <;> simp_all
+
+@[simp]
+theorem le_antisymm {e1 e2 : EffectKind} (h12: e1 ≤ e2) (h21: e2 ≤ e1) : e1 = e2 := by
+  cases e1 <;> cases e2 <;> simp_all
+
+theorem le_of_eq {e1 e2 : EffectKind} (h : e1 = e2) : e1 ≤ e2 := by
+  subst h
+  cases e1 <;> simp
+
+instance : PartialOrder EffectKind where
+  le_refl := le_refl
+  le_trans := @le_trans
+  le_antisymm := @le_antisymm
+
+end PartialOrder
+
+/-!
+### `Lattice`
+`EffectKind` forms a lattice -/
+section Lattice
+
+def sup : EffectKind → EffectKind → EffectKind
+  | .pure, .pure => .pure
+  | _, _ => .impure
+instance : Sup EffectKind := {sup}
+
+def inf : EffectKind → EffectKind → EffectKind
+  | .impure, .impure => .impure
+  | _, _ => .pure
+instance : Inf EffectKind := {inf}
+
+@[simp] theorem pure_sup_pure_eq  : pure ⊔ pure = pure    := rfl
+@[simp] theorem impure_sup_eq     : impure ⊔ e  = impure  := rfl
+@[simp] theorem sup_impure_eq     : e ⊔ impure  = impure  := by cases e <;> rfl
+
+@[simp] theorem impure_inf_impure_eq  : impure ⊓ impure = impure  := rfl
+@[simp] theorem pure_inf_eq           : pure ⊓ e = pure           := rfl
+@[simp] theorem inf_pure_eq           : e ⊓ pure = pure           := by cases e <;> rfl
+
+-- TODO: these proofs are currently quite slow, they could probablye be sped up quite a bit
+instance : Lattice EffectKind where
+  le_sup_left   := by rintro (_|_) (_|_) <;> constructor
+  le_sup_right  := by rintro (_|_) (_|_) <;> constructor
+  sup_le        := by rintro (_|_) (_|_) (_|_) <;> simp
+  inf_le_left   := by rintro (_|_) (_|_) <;> constructor
+  inf_le_right  := by rintro (_|_) (_|_) <;> constructor
+  le_inf        := by rintro (_|_) (_|_) (_|_) <;> simp
+
+end Lattice
+
+/-!
+## `liftEffect`
+-/
+
+/-- Lift a value wrapped in effect `e1` into effect `e2`, given that `e1 ≤ e2`.
+
+Said differently, this is a functor from the category of EffectKind (with `e1 ≤ e2` as its arrows)
+to Lean (with `e1.toMonad x → e2.toMonad x` as its arrows). -/
+def liftEffect [Monad m] {e1 e2 : EffectKind} {α : Type}
+    (hle : e1 ≤ e2) (v1 : e1.toMonad m α) : e2.toMonad m α :=
+  match e1, e2, hle with
+    | .pure, .pure, _ | .impure, .impure, _ => v1
+    | .pure, .impure, _ => return v1
+
+instance instMonadLiftOfLe {e1 e2 : EffectKind} (h : e1 ≤ e2) [Monad m] :
+    MonadLiftT (e1.toMonad m) (e2.toMonad m) where
+  monadLift := liftEffect h
+
+instance (eff : EffectKind) {m} [Monad m] : MonadLiftT (eff.toMonad m) m :=
+  instMonadLiftOfLe (le_impure eff)
+
+@[simp] theorem liftEffect_rfl [Monad m] (hle : eff ≤ eff) :
+    liftEffect hle (α := α) (m := m) = id := by cases eff <;> rfl
+
+@[deprecated liftEffect_rfl]
+theorem liftEffect_pure_pure [Monad m] (hle : pure ≤ pure) :
+    liftEffect hle (α := α) (m := m) = id :=
+  rfl
+
+@[simp] theorem liftEffect_pure_impure [Monad m] (hle : pure ≤ impure) :
+    liftEffect hle (α := α) (m := m) = Pure.pure :=
```

Comment:
@goens , @alexkeizer told me that there are situations where proof are not "really irrelevant", so we take the theorem as a parameter. Maybe we can make this a default argument instead?

Reply:
Oh is this about `simp` and how the proof looks like syntactically? It would be nice to understand that better, but I don't want this to block this merge. I made an issue (#253) to keep track of this one.

Reply:
Resolving conversation.

Reply:
Resolving, since we are tracking this as an issue now.